### PR TITLE
Adding election callbacks from plugin.

### PR DESCRIPTION
**Pull request by iRitwik**
Adding a public functionality to learn election results from RaftConsensus. The plugin can use it for LMP / DMP.

### CODE REVIEW

File: src/kudu/consensus/raft_consensus.cc [hunk 1]
```diff
@@ -2526,6 +2526,11 @@ RaftConfigPB RaftConsensus::CommittedConfig() const {
   return cmeta_->CommittedConfig();
 }
 
+void RaftConsensus::RegisterElectionOutcomeCallback(
+    ElectionDecisionCallback cb) {
```
Comment (anirbanr-fb) [inline src/kudu/consensus/raft_consensus.cc]: take LockGuard, since you are updating internal state

File: src/kudu/consensus/raft_consensus.cc [hunk 2]
```diff
@@ -2579,7 +2584,7 @@ void RaftConsensus::DoElectionCallback(ElectionReason reason, const ElectionResu
     LOG_WITH_PREFIX_UNLOCKED(INFO) << "Received " << election_type << " callback for term "
                                    << election_term << " while not running: "
                                    << s.ToString();
-    return;
+    return edcb_(result);
```
Comment (anirbanr-fb) [inline src/kudu/consensus/raft_consensus.cc]: this election callback will be called with the lock held. 
You should make this callback with no lock held. If this makes it more difficult,
you will somehow have to make it a tail call, like the pattern in.

void LeaderElection::CheckForDecision() 

Alternately you could wrap DoElectionCallback into DoElectionCallbackNested,
which does DoElectionCallback and then fires Nested Callback

File: src/kudu/consensus/raft_consensus.cc [hunk 3]
```diff
@@ -2678,6 +2684,19 @@ void RaftConsensus::DoElectionCallback(ElectionReason reason, const ElectionResu
...
+void RaftConsensus::NestedElectionDecisionCallback(
+    ElectionReason reason, const ElectionResult& result) {
+  DoElectionCallback(reason, result);
+  const bool was_pre_election = result.vote_request.is_pre_election();
+  UniqueLock lock(lock_);
+  const bool runClientElectionDecisionCallback =
+      (cmeta_->active_role() == RaftPeerPB::LEADER) || !was_pre_election;
+  lock.unlock();
+  if (runClientElectionDecisionCallback) {
+    edcb_(result);
```
Comment (anirbanr-fb) [inline src/kudu/consensus/raft_consensus.cc:2689]: edcb_ can be changing when RegisterElectionOutcomeCallback is called,
which is unlikely. Ideally you should not allow this callback to ever change,
and that can be ensured by setting it before Start (i.e. only in Create). The API
on consensus can give the wrong impression of its dynamic property.

File: src/kudu/consensus/raft_consensus.cc [hunk 4]
```diff
@@ -2678,6 +2684,19 @@ void RaftConsensus::DoElectionCallback(ElectionReason reason, const ElectionResu
   }
 }
 
+void RaftConsensus::NestedElectionDecisionCallback(
+    ElectionReason reason, const ElectionResult& result) {
+  DoElectionCallback(reason, result);
+  const bool was_pre_election = result.vote_request.is_pre_election();
+  UniqueLock lock(lock_);
+  const bool runClientElectionDecisionCallback =
+      (cmeta_->active_role() == RaftPeerPB::LEADER) || !was_pre_election;
```
Comment (anirbanr-fb) [inline src/kudu/consensus/raft_consensus.cc]: Is this what you intended as the logic?
it seems for true elections, you will get all notifications.
For pre-elections, you will get notifications if you are the LEADER.

I thought the logic should be && condition (notifications for successful elections only), or just plain **!was_pre_election** (all notifications for true-elections), in which you case you wont need the extra locking code.

**Review by anirbanr-fb — APPROVED** [review]
I am approving, so that you can merge when you are comfortable with it.
The 2 comments I left, both seems to be important enough to be looked into.

File: src/kudu/consensus/raft_consensus.cc (hunk 3 above)
Comment (iRitwik) [reply src/kudu/consensus/raft_consensus.cc:2689]: The other option was to use a hack via server opts. I have used that instead in this revision.

**Review by iRitwik — COMMENTED** [review]
Sorry about the condition for executing plugin callback. I got more confused when I should have just taken a break. I am requesting review again.

File: src/kudu/consensus/raft_consensus.cc (hunk 3 above)
Comment (anirbanr-fb) [reply src/kudu/consensus/raft_consensus.cc:2689]: agree

**Pull request merged**